Tasks/JavaToolInstaller/FileExtractor/JavaFilesExtractor.ts

```diff
@@ -0,0 +1,168 @@
+                .arg(file);
+            await unzip.exec(<tr.IExecOptions>{ cwd: destinationFolder });
+        }
+    }
+
+    private async sevenZipExtract(file, destinationFolder) {
+        console.log(taskLib.loc('SevenZipExtractFile', file));
+        const sevenZip = taskLib.tool(this.getSevenZipLocation());
+        sevenZip.arg('x');
+        sevenZip.arg('-o' + destinationFolder);
+        sevenZip.arg(file);
+        const execResult = sevenZip.execSync();
+        if (execResult.code != taskLib.TaskResult.Succeeded) {
+            taskLib.debug('execResult: ' + JSON.stringify(execResult));
+        }
+    }
+
+    private async tarExtract(file, destinationFolder) {
+        console.log(taskLib.loc('TarExtractFile', file));
+        const tr: tr.ToolRunner = taskLib.tool('tar');
+        tr.arg(['xzC', destinationFolder, '-f', file]);
+        tr.exec();
+    }
+
+    private extractFiles(file: string, fileEnding: string) {
+        const stats = taskLib.stats(file);
+        if (!stats) {
+            throw new Error(taskLib.loc('ExtractNonExistFile', file));
+        } else if (stats.isDirectory()) {
+            throw new Error(taskLib.loc('ExtractDirFailed', file));
+        }
+
+        if (this.win) {
+            if ('.tar' === fileEnding) { // a simple tar
+                    this.sevenZipExtract(file, this.destinationFolder);
+            } else if ('.tar.gz' === fileEnding) { // a compressed tar, e.g. 'fullFilePath/test.tar.gz'
+                    // e.g. 'fullFilePath/test.tar.gz' --> 'test.tar.gz'
+                    const shortFileName = file.substring(file.lastIndexOf(path.sep) + 1, file.length);
+                    // e.g. 'destinationFolder/_test.tar.gz_'
+                    const tempFolder = path.normalize(this.destinationFolder + path.sep + '_' + shortFileName + '_');
+                    console.log(taskLib.loc('CreateTempDir', tempFolder, file));
+
+                    // 0 create temp folder
+                    taskLib.mkdirP(tempFolder);
+
+                    // 1 extract compressed tar
+                    this.sevenZipExtract(file, tempFolder);
+                    console.log(taskLib.loc('TempDir', tempFolder));
+                    const tempTar = tempFolder + path.sep + taskLib.ls('-A', [tempFolder])[0]; // should be only one
+                    console.log(taskLib.loc('DecompressedTempTar', file, tempTar));
+                       
+                    // 2 expand extracted tar
+                    this.sevenZipExtract(tempTar, this.destinationFolder);
+
+                    // 3 cleanup temp folder
+                    console.log(taskLib.loc('RemoveTempDir', tempFolder));
+                    taskLib.rmRF(tempFolder);
+            } else if ('.zip' === fileEnding) {
+                this.unzipExtract(file, this.destinationFolder);
+            } else { // use sevenZip
+                this.sevenZipExtract(file, this.destinationFolder);
+            }
+        } else { // not windows
```

Comment:
You may want to consider putting the if-windows check inside the `.tar` and `.tar.gz` section instead of the other way around.  It may reduce duplicate code.